lib/routes/recovery-keys.js

```diff
@@ -0,0 +1,223 @@
+        }
+      },
+      handler(request, reply) {
+        log.begin('createRecoveryKey', request)
+
+        const uid = request.auth.credentials.uid
+        const sessionToken = request.auth.credentials
+        const {recoveryKeyId, recoveryData} = request.payload
+
+        customs.check(request, 'createRecoveryKey')
+          .then(createRecoveryKey)
+          .then(emitMetrics)
+          .then(() => reply({}), reply)
+
+        function createRecoveryKey() {
+          if (sessionToken.tokenVerificationId) {
+            throw errors.unverifiedSession()
+          }
+
+          return db.createRecoveryKey(uid, recoveryKeyId, recoveryData)
+        }
+
+        function emitMetrics() {
+          log.info({
+            op: 'account.recoveryKey.created',
+            uid
+          })
+
+          return request.emitMetricsEvent('recoveryKey.created', {uid})
+        }
+      }
+    },
+    {
+      method: 'GET',
+      path: '/recoveryKeys/{recoveryKeyId}',
+      config: {
+        auth: {
+          strategy: 'accountResetToken'
+        },
+        validate: {
+          params: {
+            recoveryKeyId: isA.string().max(32).required()
+          }
+        }
+      },
+      handler(request, reply) {
+        log.begin('getRecoveryKey', request)
+
+        const uid = request.auth.credentials.uid
+        const recoveryKeyId = request.params.recoveryKeyId
+        let recoveryData
+
+        customs.check(request, 'getRecoveryKey')
+          .then(getRecoveryKey)
+          .then(() => reply({recoveryData}), reply)
+
+        function getRecoveryKey() {
+          return db.getRecoveryKey(uid, recoveryKeyId)
+            .then((res) => recoveryData = res.recoveryData)
+        }
+      }
+    },
+    {
+      method: 'POST',
+      path: '/account/reset/recoveryKeys',
+      config: {
+        auth: {
+          strategy: 'accountResetToken'
+        },
+        validate: {
+          query: {
+            keys: isA.boolean().optional()
+          },
+          payload: {
+            authPW: validators.authPW,
+            wrapKb: validators.wrapKb,
+            recoveryKeyId: isA.string().max(256).regex(HEX_STRING).required()
+          }
+        }
+      },
+      handler: function (request, reply) {
+        log.begin('Password.changeRecoveryKeys', request)
+        const authPW = request.payload.authPW
+        const wrapKb = request.payload.wrapKb
+        const recoveryKeyId = request.payload.recoveryKeyId
+        const uid = request.auth.credentials.uid
+
+        let authSalt, password, verifyHash, sessionToken, keyFetchToken, account
+
+        checkRecoveryKey()
+          .then(getAccount)
+          .then(passwordReset)
+          .then(createSessionToken)
+          .then(createKeyFetchToken)
+          .then(deleteRecoveryKey)
+          .then(createResponse)
+          .then(reply, reply)
+
+        function checkRecoveryKey() {
+          return db.getRecoveryKey(uid, recoveryKeyId)
+        }
```

Comment:
I assume this is to ensure that the caller provided a correct and current recovery key, but I wonder whether it's necessary in practice.  If they're calling this API and providing wrapKb, they've almost certainly consumed a recovery key in order to get it.  Is this more of a "just in case" sort of thing?

Reply:
Correct, wanted to error on the side of caution.